packages/zql/src/ivm/flipped-join.ts

```diff
@@ -0,0 +1,440 @@
+        // the join constraint?
+        const stream = this.#parent.fetch({
+          ...req,
+          constraint: {
+            ...req.constraint,
+            ...Object.fromEntries(
+              this.#parentKey.map((key, i) => [
+                key,
+                childNode.row[this.#childKey[i]],
+              ]),
+            ),
+          },
+        });
+        const iterator = stream[Symbol.iterator]();
+        parentIterators.push(iterator);
+      }
+      const nextParentNodes: (Node | null)[] = [];
+      for (let i = 0; i < parentIterators.length; i++) {
+        const iter = parentIterators[i];
+        const result = iter.next();
+        nextParentNodes[i] = result.done ? null : result.value;
+      }
+
+      while (true) {
+        let minParentNode = null;
+        let minParentNodeChildIndexes: number[] = [];
+        for (let i = 0; i < nextParentNodes.length; i++) {
+          const parentNode = nextParentNodes[i];
+          if (parentNode === null) {
+            continue;
+          }
+          if (minParentNode === null) {
+            minParentNode = parentNode;
+            minParentNodeChildIndexes.push(i);
+          } else {
+            const compareResult =
+              this.#schema.compareRows(parentNode.row, minParentNode.row) *
+              (req.reverse ? -1 : 1);
+            if (compareResult === 0) {
+              minParentNodeChildIndexes.push(i);
+            } else if (compareResult < 0) {
+              minParentNode = parentNode;
+              minParentNodeChildIndexes = [i];
+            }
+          }
+        }
+        if (minParentNode === null) {
+          return;
+        }
+        const relatedChildNodes: Node[] = [];
+        for (const minParentNodeChildIndex of minParentNodeChildIndexes) {
+          relatedChildNodes.push(childNodes[minParentNodeChildIndex]);
+          const iter = parentIterators[minParentNodeChildIndex];
+          const result = iter.next();
+          nextParentNodes[minParentNodeChildIndex] = result.done
+            ? null
+            : result.value;
+        }
+        let overlaidRelatedChildNodes = relatedChildNodes;
+        if (
+          this.#inprogressChildChange &&
+          this.#inprogressChildChange.position &&
+          isJoinMatch(
+            this.#inprogressChildChange.change.node.row,
+            this.#childKey,
+            minParentNode.row,
+            this.#parentKey,
+          )
+        ) {
+          if (this.#inprogressChildChange.change.type === 'remove') {
+            if (
+              this.#parent
+                .getSchema()
+                .compareRows(
+                  minParentNode.row,
+                  this.#inprogressChildChange.position,
+                ) <= 0
+            ) {
+              overlaidRelatedChildNodes = relatedChildNodes.filter(
+                n => n !== this.#inprogressChildChange?.change.node,
+              );
```

Comment:
I'm a bit confused on how this relates to ` childNodes.splice(insertPos, 0, removedNode);` on line 129. Doesn't line 129 mean that `inProgressChildChange.change.node` is never in `relatedChildNodes`?